pkg/util/strategicpatch/patch.go

```diff
@@ -0,0 +1,454 @@
+		// Maybe in the future add a "concat" mode that doesn't
+		// uniqify.
+		both := append(original, patch...)
+		return uniqifyScalars(both), nil
+	}
+
+	if mergeKey == "" {
+		return nil, fmt.Errorf("cannot merge lists without merge key for type %s", elemType.Kind().String())
+	}
+
+	// First look for any special __patch elements.
+	patchWithoutSpecialElements := []interface{}{}
+	strict := false
+	for _, v := range patch {
+		typedV := v.(map[string]interface{})
+		patchType, ok := typedV[specialKey]
+		if ok {
+			if patchType == "delete" {
+				mergeValue, ok := typedV[mergeKey]
+				if ok {
+					_, originalKey, found := findMapInSliceBasedOnKeyValue(original, mergeKey, mergeValue)
+					if found {
+						// Delete the element at originalKey.
+						original = append(original[:originalKey], original[originalKey+1:]...)
+					}
+				} else {
+					return nil, fmt.Errorf("delete patch type with no merge key defined")
+				}
+			} else if patchType == "strict" {
+				strict = true
+				// Continue iterating through the array to prune any other __patch elements.
+			} else {
+				return nil, fmt.Errorf("unknown patch type found: %s", patchType)
+			}
+		} else {
+			patchWithoutSpecialElements = append(patchWithoutSpecialElements, v)
+		}
+	}
+
+	if strict {
+		return patchWithoutSpecialElements, nil
+	}
+
+	patch = patchWithoutSpecialElements
+
+	// Merge patch into original.
+	for _, v := range patch {
+		// Because earlier we confirmed that all the elements are maps.
+		typedV := v.(map[string]interface{})
+		mergeValue, ok := typedV[mergeKey]
+		if !ok {
+			return nil, fmt.Errorf("all list elements need the merge key %s", mergeKey)
+		}
+		// If we find a value with this merge key value in original, merge the
+		// maps. Otherwise append onto original.
+		originalMap, originalKey, found := findMapInSliceBasedOnKeyValue(original, mergeKey, mergeValue)
+		if found {
+			var mergedMaps interface{}
+			var err error
+			// Merge into original.
+			mergedMaps, err = mergeMap(originalMap, typedV, elemType)
+			if err != nil {
+				return nil, err
+			}
+			original[originalKey] = mergedMaps
+		} else {
+			original = append(original, v)
+		}
+	}
+
+	return original, nil
+}
+
+// This panics if any element of the slice is not a map.
+func findMapInSliceBasedOnKeyValue(m []interface{}, key string, value interface{}) (map[string]interface{}, int, bool) {
+	for k, v := range m {
+		typedV := v.(map[string]interface{})
+		valueToMatch, ok := typedV[key]
+		if ok && valueToMatch == value {
+			return typedV, k, true
+		}
+	}
+	return nil, 0, false
+}
+
+// This function takes a JSON map and sorts all the lists that should be merged
+// by key. This is needed by tests because in JSON, list order is significant,
+// but in Strategic Merge Patch, merge lists do not have significant order.
+// Sorting the lists allows for order-insensitive comparison of patched maps.
+func sortMergeListsByName(mapJSON []byte, dataStruct interface{}) ([]byte, error) {
```

Comment:
It's used in the tests.

It needs to be separate because in the tests you're comparing expected result to output. You can control how output is produced, but you can't control how someone writes their expected result (e.g. out of order), and you need to do an order-insensitive comparison between them.

When we build dry-run I'll revisit how some components of this might be able to be refactored/cleaned up.